urllib3/base.py

```diff
@@ -0,0 +1,106 @@
+    changing them as necessary. The low-level layers know how to send these
+    objects.
+    """
+    def __init__(self, method, target, headers=None, body=None):
+        #: The HTTP method in use. Must be a byte string.
+        self.method = method
+
+        #: The request target: that is, the path and query portions of the URI.
+        self.target = target
+
+        #: The request headers. These are always stored as a HTTPHeaderDict.
+        self.headers = HTTPHeaderDict(headers)
+
+        #: The request body. This is allowed to be one a few kind of objects:
+        #:    - A byte string.
+        #:    - A "readable" object.
+        #:    - An iterable of byte strings.
+        #:    - A text string (not recommended, auto-encoded to UTF-8)
+        self.body = body
+
+    def add_host(self, host, port, scheme):
+        """
+        Add the Host header, as needed.
+
+        This helper method exists to circumvent an ordering problem: the best
+        layer to add the Host header is the bottom layer, but it is the layer
+        that will add headers last. That means that they will appear at the
+        bottom of the header block.
+
+        Proxies, caches, and other intermediaries *hate* it when clients do
+        that because the Host header is routing information, and they'd like to
+        see it as early as possible. For this reason, this method ensures that
+        the Host header will be the first one emitted. It also ensures that we
+        do not duplicate the host header: if there already is one, we just use
+        that one.
+        """
+        if b'host' not in self.headers:
+            # We test against a sentinel object here to forcibly always insert
+            # the port for schemes we don't understand.
+            if port is DEFAULT_PORTS.get(scheme, object()):
+                header = host
+            else:
+                header = "{}:{}".format(host, port)
+
+            headers = HTTPHeaderDict(host=header)
+            headers._copy_from(self.headers)
+            self.headers = headers
+
+
+class Response(object):
+    """
+    The abstract low-level Response object that urllib3 works on. This is not
+    the high-level helpful Response object that is exposed at the higher layers
+    of urllib3: it's just a simple object that just exposes the lowest-level
+    HTTP semantics to allow processing by the higher levels.
+    """
+    def __init__(self, status_code, headers, body, version):
+        #: The HTTP status code of the response.
+        self.status_code = status_code
+
+        #: The headers on the response, as a HTTPHeaderDict.
+        self.headers = HTTPHeaderDict(headers)
+
+        #: The request body. This is an iterable of bytes, and *must* be
+        #: iterated if the connection is to be preserved.
+        self.body = body
+
+        #: The HTTP version of the response. Stored as a bytestring.
+        self.version = version
+
+    @property
+    def complete(self):
```

Comment:
Sorry that I haven't had a moment to pull this down yet to check, but is `body` guaranteed to have `complete`? The BytesIO objects used in the tests don't seem to. It looks like you're catching for `AttributeError` in certain places for it but not everywhere.